File: 3013-divide-an-array-into-subarrays-with-minimum-cost-ii/3013-divide-an-array-into-subarrays-with-minimum-cost-ii.java
class Solution {
    public long minimumCost(int[] nums, int k, int dist) {
        
        long result = Long.MAX_VALUE, windowSum = 0l;
        int n = nums.length;
        TreeSet<Integer> using = new TreeSet<>((a, b) -> nums[a] == nums[b] ? a - b : nums[a] - nums[b]);
        TreeSet<Integer> waiting = new TreeSet<>((a, b) -> nums[a] == nums[b] ? a - b : nums[a] - nums[b]);
        
        // Add dist + 1 num into the using set
        for(int i = 1; i <= dist + 1; i++){
            using.add(i);
            windowSum += nums[i];
        }
        // We need only k - 1 nums, move others to waiting set because we might need them in the future
        while(using.size() > k - 1){
            int i = using.pollLast();
            windowSum -= nums[i];
            waiting.add(i);
        }
        // windowSum is the sum of k - 1 nums 
        result = Math.min(result, windowSum);
        for(int i = 1; i + dist + 1 < n; i++){
            
            // window slided so add new num to the waiting set
            waiting.add(i + dist + 1);
            
            // i is the left most num in the window, will be removed from window
                
            if(using.contains(i)){
                // i is one of the k - 1 num
                // remove and update windowSum
                // poll one num from waiting set and add
                windowSum -= nums[i];
                using.remove(i);
                int j = waiting.pollFirst();
                windowSum += nums[j];
                using.add(j);
            }else{
                // i is not one of the k - 1 num, it is in the waiting set
                // remove from waiting 
                // check minimum num of the waiting set is lower than maximum num of the using set
                // If so, add to the using set, remove from waiting set
                // update window accordingly
                waiting.remove(i);
                int j1 = waiting.first(), j2 = using.last();
                if(nums[j1] < nums[j2]){
                    windowSum -= nums[j2];
                    using.remove(j2);
                    waiting.add(j2);
                    
                    windowSum += nums[j1];
                    using.add(j1);
                    waiting.remove(j1);
                }
            }
            
            result = Math.min(result, windowSum);
        }
        return result + nums[0];
    }
}

File: 3013-divide-an-array-into-subarrays-with-minimum-cost-ii/3013-divide-an-array-into-subarrays-with-minimum-cost-ii.cpp
class Solution {
    multiset<long long> l, r;
public:
    long long minimumCost(vector<int>& nums, int k, int dist) {
        int n = nums.size();
        k--;
        long long cur = nums[0];
        for (int i = 1; i <= dist + 1; i++) cur += nums[i], l.insert(nums[i]);
        while (l.size() > k) {
            cur -= *l.rbegin();
            r.insert(*l.rbegin());
            l.erase(l.find(*l.rbegin()));
        }
        long long ans = cur;

        for (int i = dist + 2; i < n; i++) {
            // Erasing an element from the set
            if (l.find(nums[i - dist - 1]) != l.end()) {
                cur -= nums[i - dist - 1];
                l.erase(l.find(nums[i - dist - 1]));
            } else {
                r.erase(r.find(nums[i - dist - 1]));
            }
            // Adding an element to the set
            if (nums[i] < *l.rbegin()) {
                cur += nums[i];
                l.insert(nums[i]);
            } else {
                r.insert(nums[i]);
            }
            // Balancing the set such that the first set has (k - 1) elements 
            while (l.size() < k) {
                cur += *r.begin();
                l.insert(*r.begin());
                r.erase(r.find(*r.begin()));
            }
            while (l.size() > k) {
                cur -= *l.rbegin();
                r.insert(*l.rbegin());
                l.erase(l.find(*l.rbegin()));
            }
            // Finally, update the minimum
            ans = min(ans, cur);
        }
        return ans;
    }
};